### PR TITLE
[Merged by Bors] - feat: some lemmas on permutation

### PR DESCRIPTION
**Pull request by AntoineChambert-Loir**
This PR prepares the way to PR #17046 and #17047 which compute the centralizer of a permutation
in a permutation group and in the alternating group.

---

- [x] depends on: #9342 
- [x] depends on: #16775
- [x] depends on: #16774 
- [x] depends on: #9602

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/GroupTheory/Perm/Centralizer.lean [hunk 1]
```diff
@@ -0,0 +1,2379 @@
+/-
+Copyright (c) 2023 Antoine Chambert-Loir. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir
+
+! This file was ported from Lean 3 source module conj_class_count
```
Comment (jcommelin) [inline Mathlib/GroupTheory/Perm/Centralizer.lean:6]: But it was never part of mathlib3, right? So I think there is not so much value in registering this in the file header.

File: Mathlib/GroupTheory/Perm/Centralizer.lean [hunk 2]
```diff
@@ -0,0 +1,2379 @@
...
+
+* We now can compute the cardinality of the set of permutations with given cycle type.
+  The condition for this cardinality to be zero is given by
+  `Equiv.Perm.card_of_cycleType_eq_zero_iff`
+  which is itself derived from `Equiv.Perm.exists_with_cycleType_iff`.
+
+* `Equiv.Perm.card_of_cycleType_mul_eq m` and `Equiv.Perm.card_of_cycleType m`
+  compute this cardinality.
+
+* `AlternatingGroup.of_cycleType_eq m`, `AlternatingGroup.of_cycleType_eq m`
```
Comment (jcommelin) [inline Mathlib/GroupTheory/Perm/Centralizer.lean:83]: This line mentions the same name twice... did you mean something else?

File: Mathlib/GroupTheory/Perm/Centralizer.lean [hunk 3]
```diff
@@ -0,0 +1,2379 @@
...
+  and `AlternatingGroup.card_of_cycleType m` give the same result
+  in the subgroup `alternatingGroup α`
+
+* `Equiv.Perm.OnCycleFactors.sign_ψ` computes the signature of the permutation induced given by `ψ`.
+
+* We finally compute on which condition the centralizer of an even permutation
+  is contained in `alternatingGroup α` and deduce the formula
+  for the cardinality of the centralizers and conjugacy classes
```
Comment (jcommelin) [inline Mathlib/GroupTheory/Perm/Centralizer.lean:91]: What are the names of the results where this is done?

File: Mathlib/GroupTheory/Perm/Centralizer.lean (hunk 1 above)
Comment (jcommelin) [reply Mathlib/GroupTheory/Perm/Centralizer.lean:6]: And maybe this also means that the `#align` statements are not needed?

Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/Perm/Centralizer.lean:6]: It was provided by mathport — I'll delete them.

File: Mathlib/GroupTheory/Perm/Centralizer.lean (hunk 2 above)
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/Perm/Centralizer.lean:83]: no, it's a typo.

File: Mathlib/GroupTheory/Perm/Centralizer.lean (hunk 3 above)
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/Perm/Centralizer.lean:91]: I adjusted the docstring. It seems the last lemma is an unfilled promise.

**Comment by jcommelin** [comment]
I pushed a bunch of golfing and cleanup of the first ~600 lines to `jmc-ConjClassCount`. I hope the diff is useful to you.

I think those first ~600 lines of this PR could make a good standalone PR.

**Comment by AntoineChambert-Loir** [comment]
> I pushed a bunch of golfing and cleanup of the first ~600 lines to `jmc-ConjClassCount`. I hope the diff is useful to you.

Thank you Johann! 
How do you suggest I work with it? make a local copy and adjust my file? merge the two branches?

**Comment by jcommelin** [comment]
@AntoineChambert-Loir Both could work. The merge should be rather straightforward, although it might conflict with the removal of the `#align` statements.
You can also cherry-pick changes by hand.

**Comment by riccardobrasca** [comment]
What about putting the first lemmas (up to line 600 or something) to a new file `Mathlib.GroupTheory.Perm.Lemmas` or something like that (and in a separate PR...)?

**Comment by AntoineChambert-Loir** [comment]
Yes, I'll do that (and probably find them adequate place in the GroupTheory.Perm hierarchy)

**Review by riccardobrasca — COMMENTED** [review]
I know I always say the same thing, but it seems to me there is an easy way of splitting the PR:
* you modify three files
* new file 1
* new file 2

What do you think?

File: Mathlib/GroupTheory/SpecificGroups/Alternating/Centralizer.lean [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+theorem sameCycle_iff_cycleOf_eq_of_mem_support
+    {g : Perm α} {x y : α} (hx : x ∈ g.support) (hy : y ∈ g.support) :
+    g.SameCycle x y ↔ g.cycleOf x = g.cycleOf y := by
+  constructor
+  · exact SameCycle.cycleOf_eq
+  · intro h
+    rw [← mem_support_cycleOf_iff' (mem_support.mp hx), h]
+    rw [mem_support_cycleOf_iff' (mem_support.mp hy)]
+
+/- theorem cycleType_prime_order' (σ : Equiv.Perm α) (p : ℕ) [hp : Fact (Nat.Prime p)] :
```
[the remark this answers is not in the document]
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/SpecificGroups/Alternating/Centralizer.lean:52]: Well, it is mathematically equivalent to the theorem just below, but I wasn't sure what form was preferable in mathlib. The one I kept is clearer, but needs more work to know something about `σ.cycleType`. That one allows to rewrite `σ.cycleType` in a formula.

**Comment by AntoineChambert-Loir** [comment]
@riccardobrasca : I just did what you suggested, there are now two subsequent PR, #17046 and #17047 .

**Review by riccardobrasca — APPROVED** [review]
Thanks!

bors d+

**Comment by riccardobrasca** [comment]
@AntoineChambert-Loir this is almost ready to go. Can you please take into account the last comments? Thanks!

File: Mathlib/GroupTheory/Perm/Cycle/Factors.lean [hunk 5]
```diff
@@ -502,6 +503,12 @@ theorem cycleOf_mem_cycleFactorsFinset_iff {f : Perm α} {x : α} :
     · rw [cycleOf_apply_of_not_sameCycle H] at hy
       contradiction
 
+lemma cycleOf_ne_one_iff_mem_cycleFactorsFinset (g : Equiv.Perm α) {x : α} :
```
[the remark this answers is not in the document]
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/Perm/Cycle/Factors.lean:505]: You're right, iff lemmas usually have all of their variables implicit.

**Pull request closed without merging**